src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RSAAndroid.cs

```diff
@@ -0,0 +1,846 @@
+                        parameters.InverseQ != null)
+                    {
+                        return false;
+                    }
+                }
+                else
+                {
+                    if (parameters.P == null ||
+                        parameters.DP == null ||
+                        parameters.Q == null ||
+                        parameters.DQ == null ||
+                        parameters.InverseQ == null)
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private void ThrowIfDisposed()
+            {
+                if (_key == null)
+                {
+                    throw new ObjectDisposedException(nameof(RSA));
+                }
+            }
+
+            private SafeRsaHandle GetKey()
+            {
+                ThrowIfDisposed();
+
+                SafeRsaHandle key = _key.Value;
+
+                if (key == null || key.IsInvalid)
+                {
+                    throw new CryptographicException(SR.Cryptography_OpenInvalidHandle);
+                }
+
+                return key;
+            }
+
+            private static void CheckReturn(int returnValue)
+            {
+                if (returnValue == -1)
+                {
+                    throw new CryptographicException();
+                }
+            }
+
+            private static void CheckBoolReturn(int returnValue)
+            {
+                if (returnValue != 1)
+                {
+                throw new CryptographicException();
+                }
+            }
+
+            private SafeRsaHandle GenerateKey()
+            {
+                SafeRsaHandle key = Interop.AndroidCrypto.RsaCreate();
+                bool generated = false;
+
+                if (key is null || key.IsInvalid)
+                {
+                    throw new CryptographicException();
+                }
+
+                try
+                {
+                    using (SafeBignumHandle exponent = Interop.Crypto.CreateBignum(DefaultExponent))
+                    {
+                        // The documentation for RSA_generate_key_ex does not say that it returns only
+                        // 0 or 1, so the call marshals it back as a full Int32 and checks for a value
+                        // of 1 explicitly.
+                        int response = Interop.AndroidCrypto.RsaGenerateKeyEx(
+                            key,
+                            KeySize,
+                            exponent);
+
+                        CheckBoolReturn(response);
+                        generated = true;
+                    }
+                }
+                finally
+                {
+                    if (!generated)
+                    {
+                        key.Dispose();
+                    }
+                }
+
+                return key;
+            }
+
```

Comment:
I have mixed feelings... but I'll say it anyways: "We could route SignData and VerifyData into the signature area that Java/Android exposes and avoid needing to use our manual padding routines when the -Data members are used."

Usage says that SignData and VerifyData are about 3 times more likely to be called than SignHash and VerifyHash.

Reply:
Yes, I thought about doing that. But I feel like since we already need to have the code path that does the hashing and sign/verify steps individually, I don't see a good reason to add a separate code path just to call into the Java code to do the same thing, unless there's some reason why it's more secure for us to pass down to Java/Android in those specific cases.

We'd still need to manually handle the PSS padding in all cases since that isn't available on Android on all the API levels we support. As a result, we could only move over the data+pkcs#1 combo to Java/Android, which just doesn't feel worth it to me.

Reply:
OK.  If we could make both RSASSA-PKCS1 and RSASSA-PSS route through it always then I'd still be on the fence, but at PKCS1-only I agree it's not worth it at this time.